MediaBrowser.Controller/MediaEncoding/TranscodingSegmentCleaner.cs

```diff
@@ -0,0 +1,178 @@
+    protected virtual void Dispose(bool disposing)
+    {
+        if (disposing)
+        {
+            DisposeTimer();
+        }
+    }
+
+    private EncodingOptions GetOptions()
+    {
+        return _config.GetEncodingOptions();
+    }
+
+    private async void TimerCallback(object? state)
+    {
+        if (_job.HasExited)
+        {
+            DisposeTimer();
+            return;
+        }
+
+        var options = GetOptions();
+        var enableSegmentDeletion = options.EnableSegmentDeletion;
+        var segmentKeepSeconds = Math.Max(options.SegmentKeepSeconds, 20);
+
+        if (enableSegmentDeletion)
+        {
+            var downloadPositionTicks = _job.DownloadPositionTicks ?? 0;
+            var downloadPositionSeconds = Convert.ToInt64(TimeSpan.FromTicks(downloadPositionTicks).TotalSeconds);
+
+            if (downloadPositionSeconds > 0 && segmentKeepSeconds > 0 && downloadPositionSeconds > segmentKeepSeconds)
+            {
+                var idxMaxToDelete = (downloadPositionSeconds - segmentKeepSeconds) / _segmentLength;
+
+                if (idxMaxToDelete > 0)
+                {
+                    await DeleteSegmentFiles(_job, 0, idxMaxToDelete, 1500).ConfigureAwait(false);
+                }
+            }
+        }
+    }
+
+    private async Task DeleteSegmentFiles(TranscodingJob job, long idxMin, long idxMax, int delayMs)
+    {
+        var path = job.Path ?? throw new ArgumentException("Path can't be null.");
+
+        _logger.LogDebug("Deleting segment file(s) index {Min} to {Max} from {Path}", idxMin, idxMax, path);
+
+        await Task.Delay(delayMs).ConfigureAwait(false);
+
+        try
+        {
+            if (job.Type == TranscodingJobType.Hls)
+            {
+                DeleteHlsSegmentFiles(path, idxMin, idxMax);
+            }
+        }
+        catch (Exception ex)
+        {
+            _logger.LogDebug(ex, "Error deleting segment file(s) {Path}", path);
+        }
+    }
+
+    private void DeleteHlsSegmentFiles(string outputFilePath, long idxMin, long idxMax)
+    {
+        var directory = Path.GetDirectoryName(outputFilePath)
+                        ?? throw new ArgumentException("Path can't be a root directory.", nameof(outputFilePath));
+
+        var name = Path.GetFileNameWithoutExtension(outputFilePath);
+
+        var filesToDelete = _fileSystem.GetFilePaths(directory)
+            .Where(f => long.TryParse(Path.GetFileNameWithoutExtension(f).Replace(name, string.Empty, StringComparison.Ordinal), out var idx)
+                        && (idx >= idxMin && idx <= idxMax));
+
+        List<Exception>? exs = null;
+        foreach (var file in filesToDelete)
+        {
+            try
+            {
+                _logger.LogDebug("Deleting HLS segment file {0}", file);
+                _fileSystem.DeleteFile(file);
+            }
+            catch (IOException ex)
+            {
+                (exs ??= new List<Exception>(4)).Add(ex);
```

Comment:
Why 4?

Reply:
I was also wondering why 4. Because the code here is basically taken from TranscodeManager. I can remove it later.

https://github.com/jellyfin/jellyfin/blob/b6d130ae2d337294904f6d46d943e9f82c162978/MediaBrowser.MediaEncoding/Transcoding/TranscodeManager.cs#L324

Reply:
Strange. I would just remove it.

Reply:
https://github.com/search?q=repo%3Ajellyfin%2Fjellyfin+%3E%284&type=code

Similar situations occur elsewhere, I just don't get it. Do I need to clean them up in this PR?

Reply:
Microoptimization by me, by default list gets created with an empty internal array, when the add method gets called the fast path is designed when there is enough space for the item, otherwise another method is called (not inlined) that grows the array. All of this can be skipped by creating the list with the first default capacity of 4.
Not necessary at all, just something I added when I wrote the code is is copied from.

Reply:
I also speculate that this may be an optimization for the CLR. So i didn't touch it.

Reply:
@nyanmisaka this _used_ to be a known optimization for Dotnet Framework but ever since Framework 4.7.2 the clr got very good at predicating the initial size of list arrays. Also we are nowhere near the performance or amount of arrays where this would be justified.